doc/cookbook/testing/Testing.lhs

```diff
@@ -0,0 +1,456 @@
+
+So we've got an Elasticsearch server and a client to talk to it. Let's now
+build a simple app server that uses this client to retrieve documents. This
+is somewhat contrived, but hopefully it illustrates the typical three-tier
+application architecture.
+
+One note: we're also going to take advantage of `lens-aeson` here, which may
+look a bit foreign. The gist of it is that we're going to traverse a JSON
+`Value` from Elasticsearch and try to extract some  kind of document to
+return.
+
+Imagine, then, that this is our real server implementation:
+
+```haskell
+type DocApi =
+  "docs" :> Capture "docId" Integer :> Get '[JSON] Value
+
+docServer :: Text -> Text -> Server DocApi
+docServer esHost esPort = getDocById esHost esPort
+
+-- Our Handler tries to get a doc from Elasticsearch and then tries to parse
+-- it. Unfortunately, there's a lot of opportunity for failure in these
+-- actions
+getDocById :: Text -> Text -> Integer -> Handler Value
+getDocById esHost esPort docId = do
+  -- Our Servant Client function returns Either ServantError Value here:
+  docRes <- liftIO $ runSearchClient esHost esPort (getDocument docId)
+  case docRes of
+    Left err -> throwError $ err404 { errBody = "Failed looking up content" }
+    Right value -> do
+      -- we'll either fail to parse our document or we'll return it
+      case value ^? _Object . ix "_source" of
+        Nothing -> throwError $ err400 { errBody = "Failed parsing content" }
+        Just obj -> pure obj
+```
+
+### Testing Our Backend
+
+So the above represents our application and is close to a server we may
+actually deploy. How then shall we test this application?
+
+Ideally, we'd like it to make requests of a *real* Elasticsearch server, but
+we certainly don't want our tests to trigger requests to a live, production
+database. In addition, we don't want to depend on our real Elasticsearch
+server having specific, consistent results for us to test against, because
+that would make our tests flaky (and flaky tests are sometimes described as
+worse than not having tests at all).
+
+One solution to this is to create a trivial Elasticsearch server as part of
+our testing code. We can do this relatively easily because we already have
+an API definition for it above. With a *real* server, we can then let our own
+application make requests of it and we'll simulate different scenarios in
+order to make sure our application responds the way we expect it to.
+
+Let's start with some helpers which will allow us to run a testing version
+of our Elasticsearch server in another thread:
+
+```haskell
+-- | We'll run the Elasticsearch server so we can test behaviors
+withElasticsearch :: IO () -> IO ()
+withElasticsearch action =
+  bracket (liftIO $ C.forkIO $ Warp.run 9999 esTestApp)
+    C.killThread
+    (const action)
+
+esTestApp :: Application
+esTestApp = serve (Proxy :: Proxy SearchAPI) esTestServer
+
+esTestServer :: Server SearchAPI
+esTestServer = getESDocument
+
```

Comment:
should include a comment here probably about how this is a *mock* representation of a handler for our ES server.